JavaKotlinMavenInterop/kotlin-lib/src/main/kotlin/com/mongodb/devicesync/kotlin/RealmRepository.kt

```diff
@@ -0,0 +1,102 @@
+package com.mongodb.devicesync.kotlin
+
+import io.reactivex.rxjava3.core.Observable
+import io.realm.kotlin.MutableRealm
+import io.realm.kotlin.Realm
+import io.realm.kotlin.RealmConfiguration
+import io.realm.kotlin.ext.query
+import io.realm.kotlin.notifications.ResultsChange
+import io.realm.kotlin.types.BaseRealmObject
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.rx3.asObservable
+import java.util.concurrent.Executors
+
+class CancellationToken(val job: Job) {
+    fun cancel() {
+        job.cancel()
+    }
+}
+
+class RealmRepository {
+
+    interface UpdateCallback {
+        fun update(realm: MutableRealm)
+    }
+
+    interface EventCallback<E: BaseRealmObject> {
+        fun update(item: ResultsChange<E>)
+    }
+
+    lateinit var realm: Realm
+    val realmDispatchers = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() - 1).asCoroutineDispatcher()
+    val realmScope: CoroutineScope = CoroutineScope(CoroutineName("RealmScope") + realmDispatchers)
+
+    fun openRealm() {
+        val config = RealmConfiguration.Builder(schema = setOf(Person::class, Child::class))
+            .name("javainterop-example.realm")
+            .build()
+        Realm.deleteRealm(config) // Cleanup any old test data
+        realm = Realm.open(config)
+    }
+
+    fun closeRealm() {
+        checkRealm()
+        realmScope.cancel("Closing Realm")
+        realmDispatchers.close()
+        realm.close()
+    }
+
+    fun writeData(person: Person): Person {
+        return realm.writeBlocking {
+            copyToRealm(person)
+        }
+    }
+
+    fun writeData(list: List<Person>) {
+        realm.writeBlocking {
+            list.forEach { person: Person ->
+                copyToRealm(person)
+            }
+        }
+    }
+
+    fun updateData(callback: UpdateCallback) {
+        realm.writeBlocking {
+            callback.update(this)
+        }
+    }
+
+    fun readData(query: String): List<Person> {
+        return realm.query<Person>(query).find()
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    fun updatesAsRxJavaObserverable(): Observable<ResultsChange<Person>> {
+        return realm.query<Person>("TRUEPREDICATE").asFlow().asObservable()
+    }
+
+    // Expose as callbacks with a cancellation token
+    fun updatesAsCallbacks(callback: EventCallback<Person>): CancellationToken {
+        val job: Job = realmScope.launch {
+            realm.query<Person>("TRUEPREDICATE").asFlow()
+                .collect { event: ResultsChange<Person> ->
+                    callback.update(event)
+                }
+        }
+        return CancellationToken(job)
```

Comment:
You could also return Java's [Closeable](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html) instead of `CancellationToken`